### ISSUE TITLE
After #1696 I now get a totally garbage image using qt6 frontend:

### ISSUE THREAD

**Issue by 375gnu** [CelestiaProject/Celestia#1759]
After #1696 I now get a totally garbage image using qt6 frontend:
![test](https://github.com/CelestiaProject/Celestia/assets/17218209/396d7434-bbf3-4fe8-87b3-aea88e1598c8)

gtk frontend still works fine.
Running celestia-qt6 from CLI doesn't show any wrong message, so I'm not sure how to report more useful information.

_Originally posted by @mattiaverga in https://github.com/CelestiaProject/Celestia/issues/1461#issuecomment-1621292138_

**Comment by munix9** [comment]
works fine on openSUSE TW
![test](https://github.com/CelestiaProject/Celestia/assets/44939650/4e6c96f0-113b-46e0-bf7a-fe9d1a04ae7b)
![grafik](https://github.com/CelestiaProject/Celestia/assets/44939650/5d00cb38-ea3f-40ec-97db-04e0ed96fbb6)

**Comment by mattiaverga** [comment]
Could be hardware related?
Adding info as much as possible:

package built with
```
%cmake -DENABLE_DATA=ON \
       -DENABLE_QT5=OFF \
       -DENABLE_QT6=ON \
       -DENABLE_GTK=ON \
       -DENABLE_FFMPEG=ON \
       -DENABLE_LIBAVIF=ON \
       -DUSE_WAYLAND=ON \
       -DGIT_COMMIT="%{shortcommit}"
```
build logs available at
https://download.copr.fedorainfracloud.org/results/mattia/Astronomy/fedora-38-x86_64/06145695-celestia/

OpenGL info:
```
OpenGL version: 4.6 (Compatibility Profile) Mesa 23.1.2
Vendor: AMD
Renderer: AMD Radeon Graphics (renoir, LLVM 16.0.5, DRM 3.52, 6.3.8-200.fc38.x86_64)
GLSL Version: 4.60
Max simultaneous textures: 192
Maximum texture size: 16384
Point size range: 0 - 2048
Point size granularity: 0.12
Number of interpolators: 128
Max anisotropy filtering: 16.00

Supported extensions:
GL_ARB_multisample
GL_EXT_abgr
GL_EXT_bgra
GL_EXT_blend_color
GL_EXT_blend_minmax
GL_EXT_blend_subtract
GL_EXT_copy_texture
GL_EXT_subtexture
GL_EXT_texture_object
GL_EXT_vertex_array
GL_EXT_compiled_vertex_array
GL_EXT_texture
GL_EXT_texture3D
GL_IBM_rasterpos_clip
GL_ARB_point_parameters
GL_EXT_draw_range_elements
GL_EXT_packed_pixels
GL_EXT_point_parameters
GL_EXT_rescale_normal
GL_EXT_separate_specular_color
GL_EXT_texture_edge_clamp
GL_SGIS_generate_mipmap
GL_SGIS_texture_border_clamp
GL_SGIS_texture_edge_clamp
GL_SGIS_texture_lod
GL_ARB_framebuffer_sRGB
GL_ARB_multitexture
GL_EXT_framebuffer_sRGB
GL_IBM_multimode_draw_arrays
GL_IBM_texture_mirrored_repeat
GL_ARB_texture_cube_map
GL_ARB_texture_env_add
GL_ARB_transpose_matrix
GL_EXT_blend_func_separate
GL_EXT_fog_coord
GL_EXT_multi_draw_arrays
GL_EXT_secondary_color
GL_EXT_texture_env_add
GL_EXT_texture_filter_anisotropic
GL_EXT_texture_lod_bias
GL_INGR_blend_func_separate
GL_NV_blend_square
GL_NV_light_max_exponent
GL_NV_texgen_reflection
GL_NV_texture_env_combine4
GL_S3_s3tc
GL_SUN_multi_draw_arrays
GL_ARB_texture_border_clamp
GL_ARB_texture_compression
GL_EXT_framebuffer_object
GL_EXT_texture_compression_s3tc
GL_EXT_texture_env_combine
GL_EXT_texture_env_dot3
GL_MESA_window_pos
GL_NV_packed_depth_stencil
GL_NV_texture_rectangle
GL_ARB_depth_texture
GL_ARB_occlusion_query
GL_ARB_shadow
GL_ARB_texture_env_combine
GL_ARB_texture_env_crossbar
GL_ARB_texture_env_dot3
GL_ARB_texture_mirrored_repeat
GL_ARB_window_pos
GL_ATI_fragment_shader
GL_EXT_stencil_two_side
GL_EXT_texture_cube_map
GL_NV_copy_depth_to_color
GL_NV_depth_clamp
GL_NV_fog_distance
GL_NV_half_float
GL_APPLE_packed_pixels
GL_ARB_draw_buffers
GL_ARB_fragment_program
GL_ARB_fragment_shader
GL_ARB_shader_objects
GL_ARB_vertex_program
GL_ARB_vertex_shader
GL_ATI_draw_buffers
GL_ATI_texture_env_combine3
GL_ATI_texture_float
GL_EXT_depth_bounds_test
GL_EXT_shadow_funcs
GL_EXT_stencil_wrap
GL_MESA_pack_invert
GL_NV_primitive_restart
GL_ARB_depth_clamp
GL_ARB_fragment_program_shadow
GL_ARB_half_float_pixel
GL_ARB_occlusion_query2
GL_ARB_point_sprite
GL_ARB_shading_language_100
GL_ARB_sync
GL_ARB_texture_non_power_of_two
GL_ARB_vertex_buffer_object
GL_ATI_blend_equation_separate
GL_EXT_blend_equation_separate
GL_OES_read_format
GL_ARB_color_buffer_float
GL_ARB_pixel_buffer_object
GL_ARB_texture_compression_rgtc
GL_ARB_texture_float
GL_ARB_texture_rectangle
GL_ATI_texture_compression_3dc
GL_EXT_packed_float
GL_EXT_pixel_buffer_object
GL_EXT_texture_compression_dxt1
GL_EXT_texture_compression_rgtc
GL_EXT_texture_mirror_clamp
GL_EXT_texture_rectangle
GL_EXT_texture_sRGB
GL_EXT_texture_shared_exponent
GL_ARB_framebuffer_object
GL_EXT_framebuffer_blit
GL_EXT_framebuffer_multisample
GL_EXT_packed_depth_stencil
GL_ARB_vertex_array_object
GL_ATI_separate_stencil
GL_ATI_texture_mirror_once
GL_EXT_draw_buffers2
GL_EXT_draw_instanced
GL_EXT_gpu_program_parameters
GL_EXT_gpu_shader4
GL_EXT_texture_array
GL_EXT_texture_compression_latc
GL_EXT_texture_integer
GL_EXT_texture_sRGB_decode
GL_EXT_timer_query
GL_OES_EGL_image
GL_AMD_performance_monitor
GL_EXT_texture_buffer_object
GL_AMD_texture_texture4
GL_ARB_copy_buffer
GL_ARB_depth_buffer_float
GL_ARB_draw_instanced
GL_ARB_half_float_vertex
GL_ARB_instanced_arrays
GL_ARB_map_buffer_range
GL_ARB_texture_buffer_object
GL_ARB_texture_rg
GL_ARB_texture_swizzle
GL_ARB_vertex_array_bgra
GL_EXT_texture_swizzle
GL_EXT_vertex_array_bgra
GL_NV_conditional_render
GL_AMD_conservative_depth
GL_AMD_depth_clamp_separate
GL_AMD_draw_buffers_blend
GL_AMD_seamless_cubemap_per_texture
GL_AMD_shader_stencil_export
GL_ARB_ES2_compatibility
GL_ARB_blend_func_extended
GL_ARB_compatibility
GL_ARB_debug_output
GL_ARB_draw_buffers_blend
GL_ARB_draw_elements_base_vertex
GL_ARB_explicit_attrib_location
GL_ARB_fragment_coord_conventions
GL_ARB_provoking_vertex
GL_ARB_sample_shading
GL_ARB_sampler_objects
GL_ARB_seamless_cube_map
GL_ARB_shader_stencil_export
GL_ARB_shader_texture_lod
GL_ARB_tessellation_shader
GL_ARB_texture_buffer_object_rgb32
GL_ARB_texture_cube_map_array
GL_ARB_texture_gather
GL_ARB_texture_multisample
GL_ARB_texture_query_lod
GL_ARB_texture_rgb10_a2ui
GL_ARB_uniform_buffer_object
GL_ARB_vertex_type_2_10_10_10_rev
GL_ATI_meminfo
GL_EXT_provoking_vertex
GL_EXT_texture_snorm
GL_MESA_texture_signed_rgba
GL_NV_copy_image
GL_NV_texture_barrier
GL_ARB_draw_indirect
GL_ARB_get_program_binary
GL_ARB_gpu_shader5
GL_ARB_gpu_shader_fp64
GL_ARB_robustness
GL_ARB_separate_shader_objects
GL_ARB_shader_bit_encoding
GL_ARB_shader_precision
GL_ARB_shader_subroutine
GL_ARB_texture_compression_bptc
GL_ARB_timer_query
GL_ARB_transform_feedback2
GL_ARB_transform_feedback3
GL_ARB_vertex_attrib_64bit
GL_ARB_viewport_array
GL_EXT_direct_state_access
GL_EXT_shader_image_load_store
GL_EXT_vertex_attrib_64bit
GL_NV_vdpau_interop
GL_AMD_multi_draw_indirect
GL_ANGLE_texture_compression_dxt3
GL_ANGLE_texture_compression_dxt5
GL_ARB_base_instance
GL_ARB_compressed_texture_pixel_storage
GL_ARB_conservative_depth
GL_ARB_internalformat_query
GL_ARB_map_buffer_alignment
GL_ARB_shader_atomic_counters
GL_ARB_shader_image_load_store
GL_ARB_shading_language_420pack
GL_ARB_shading_language_packing
GL_ARB_texture_storage
GL_ARB_transform_feedback_instanced
GL_EXT_framebuffer_multisample_blit_scaled
GL_EXT_transform_feedback
GL_AMD_query_buffer_object
GL_AMD_shader_trinary_minmax
GL_AMD_vertex_shader_layer
GL_AMD_vertex_shader_viewport_index
GL_ARB_ES3_compatibility
GL_ARB_arrays_of_arrays
GL_ARB_clear_buffer_object
GL_ARB_compute_shader
GL_ARB_copy_image
GL_ARB_explicit_uniform_location
GL_ARB_fragment_layer_viewport
GL_ARB_framebuffer_no_attachments
GL_ARB_invalidate_subdata
GL_ARB_multi_draw_indirect
GL_ARB_program_interface_query
GL_ARB_robust_buffer_access_behavior
GL_ARB_shader_image_size
GL_ARB_shader_storage_buffer_object
GL_ARB_stencil_texturing
GL_ARB_texture_buffer_range
GL_ARB_texture_query_levels
GL_ARB_texture_storage_multisample
GL_ARB_texture_view
GL_ARB_vertex_attrib_binding
GL_KHR_debug
GL_KHR_robustness
GL_KHR_texture_compression_astc_ldr
GL_AMD_pinned_memory
GL_ARB_bindless_texture
GL_ARB_buffer_storage
GL_ARB_clear_texture
GL_ARB_compute_variable_group_size
GL_ARB_enhanced_layouts
GL_ARB_indirect_parameters
GL_ARB_internalformat_query2
GL_ARB_multi_bind
GL_ARB_query_buffer_object
GL_ARB_seamless_cubemap_per_texture
GL_ARB_shader_draw_parameters
GL_ARB_shader_group_vote
GL_ARB_shading_language_include
GL_ARB_sparse_texture
GL_ARB_texture_mirror_clamp_to_edge
GL_ARB_texture_stencil8
GL_ARB_vertex_type_10f_11f_11f_rev
GL_EXT_debug_label
GL_EXT_shader_integer_mix
GL_NVX_gpu_memory_info
GL_ARB_ES3_1_compatibility
GL_ARB_clip_control
GL_ARB_conditional_render_inverted
GL_ARB_cull_distance
GL_ARB_derivative_control
GL_ARB_direct_state_access
GL_ARB_get_texture_sub_image
GL_ARB_pipeline_statistics_query
GL_ARB_shader_texture_image_samples
GL_ARB_sparse_buffer
GL_ARB_texture_barrier
GL_ARB_transform_feedback_overflow_query
GL_EXT_polygon_offset_clamp
GL_EXT_shader_image_load_formatted
GL_KHR_blend_equation_advanced
GL_KHR_context_flush_control
GL_KHR_robust_buffer_access_behavior
GL_NV_shader_atomic_int64
GL_ARB_ES3_2_compatibility
GL_ARB_gpu_shader_int64
GL_ARB_parallel_shader_compile
GL_ARB_shader_atomic_counter_ops
GL_ARB_shader_ballot
GL_ARB_shader_clock
GL_ARB_shader_viewport_layer_array
GL_ARB_sparse_texture2
GL_ARB_sparse_texture_clamp
GL_EXT_shader_samples_identical
GL_EXT_texture_sRGB_R8
GL_KHR_no_error
GL_KHR_texture_compression_astc_sliced_3d
GL_ARB_gl_spirv
GL_ARB_spirv_extensions
GL_EXT_window_rectangles
GL_MESA_shader_integer_functions
GL_ARB_polygon_offset_clamp
GL_ARB_texture_filter_anisotropic
GL_EXT_memory_object
GL_EXT_memory_object_fd
GL_EXT_semaphore
GL_EXT_semaphore_fd
GL_KHR_parallel_shader_compile
GL_NV_alpha_to_coverage_dither_control
GL_AMD_framebuffer_multisample_advanced
GL_EXT_EGL_image_storage
GL_EXT_texture_shadow_lod
GL_INTEL_blackhole_render
GL_MESA_framebuffer_flip_y
GL_NV_compute_shader_derivatives
GL_EXT_EGL_sync
GL_EXT_demote_to_helper_invocation
GL_NV_ES1_1_compatibility
```

![Screenshot_20230706_092803](https://github.com/CelestiaProject/Celestia/assets/17218209/b9a7014e-1c46-40db-818a-d4a613e4edd4)

**Comment by mattiaverga** [comment]
Found the problem: I have this `.celestia.cfg` in my home:
```
Configuration
{
    AntialiasingSamples        4
}
```
Removing it produces good screenshot both in qt6 and gtk frontend!
So, qt6 frontend doesn't like antialiasing settings ...

**Comment by 375gnu** [comment]
That's weird.

**Comment by munix9** [comment]
I can confirm this: With e.g. AntialiasingSamples 4 in both qt5 and qt6 I get garbage as an image.
They look a bit different, but in both cases they are unusable pixel clutter.

**Comment by ajtribick** [comment]
Is this the usual OpenGL issue that `glReadPixels` doesn't work on multisampled framebuffers, and you have to transfer it to a single sampled framebuffer instead (e.g. with `glBlitFramebuffer`, or whatever the equivalent is for OpenGL ES 2.0)

**Comment by dave-kaye** [comment]
Utilizing openSUSE Leap 15.2 and the 07062023 git version 062d5b2 of Celestia compiled with openGL where ENABLE_GLES=OFF and USE_WAYLAND=OFF.

On both Qt5 and GTK3 versions of Celestia, I ran antialiasingsamples at both 1 and 4 samples. Unlike munix9, all settings gave good images.

Screenshots from about -

![about](https://github.com/CelestiaProject/Celestia/assets/91984396/39f67f78-8657-4302-80e1-a5db311e4c2b)


![opengl](https://github.com/CelestiaProject/Celestia/assets/91984396/ee60c8b6-6ab5-4d69-b12a-1f03eddaf389)



Then I recompiled for only the Qt5 version with openGL where ENABLE_GLES=ON and USE_WAYLAND=OFF.  The image also was good with anti-aliasing samples set to 4.  I am unable to compile with wayland due to header file issues that I would have to fix.

Just to verify that the anti-aliasing settings were what I expected, I inserted the following code into qtqppwin.cpp near line 1300.  [Using munix9 about box extension code so line number is approximate - showing neighboring code for location.]

    if (info.count("MaxAnisotropy") > 0)
    {
        out << QString(_("<b>Max anisotropy filtering:</b> %1")).arg(info["MaxAnisotropy"].c_str());
        out << "<br>\n";
    }

    // Show configuration anti-aliasing sample size
    out << QString(_("<b>Anti-aliasing samples:</b> %1")).arg(m_appCore->getConfig()->aaSamples);
    out << "<br>\n";    
    
    out << "<br>\n";

    // Show all supported extensions
    if (info.count("Extensions") > 0)

This adds the anti-aliasing sample size in the Qt help -> openGL info window.

p.s. The mirroring patch from issue #1696 works for me and cleared my problem.  THANKS!

**Comment by ajtribick** [comment]
Could the original issue have been due to calling `glPixelStorei(GL_PACK_INVERT_MESA, GL_TRUE);` if the `MESA_pack_invert` extension is detected?

Maybe worth going back to the `glWidget->grabFramebuffer()` and avoiding using the extension if the front-end is Qt (because presumably the widget's code doesn't expect us to invert the row order).

**Comment by mattiaverga** [comment]
> Could the original issue have been due to calling `glPixelStorei(GL_PACK_INVERT_MESA, GL_TRUE);` if the `MESA_pack_invert` extension is detected?
> 
> https://github.com/CelestiaProject/Celestia/blob/062d5b254a5ee8bacb7cc71a36fb6aa1950d09b9/src/celengine/render.cpp#L555-L558
> 
> Maybe worth going back to the `glWidget->grabFramebuffer()` and avoiding using the extension if the front-end is Qt (because presumably the widget's code doesn't expect us to invert the row order).

I've done a rebuild patched to revert #1696 and removing the above lines from code (as I don't know how to conditionalize for only non-qt interface). This way everything works fine (except, of course, the gtk screenshot is now mirrored).
